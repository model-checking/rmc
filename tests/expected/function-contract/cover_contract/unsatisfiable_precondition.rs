// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
// kani-flags: -Zfunction-contracts

// Test `cover_contract` functionality, which fails verification for an unsatisfiable precondition
// or unreachable postcondition.
// See https://github.com/model-checking/kani/issues/2793

// Test that verification fails for unsatisfiable preconditions.

// Unsatisfiable precondition; separate requires clauses.
// The postcondition is unreachable because of the unsatisfiable precondition.
#[kani::requires(a > 5)]
#[kani::requires(a < 4)]
#[kani::ensures(|result: &u32| *result == a)]
fn separate_requires(a: u32) -> u32 {
    panic!("This code is never tested")
}

#[kani::proof_for_contract(separate_requires)]
fn prove_separate_requires() {
    let x: u32 = kani::any();
    separate_requires(x);
}

// Unsatisfiable precondition; single requires clause
// The postcondition is unreachable because of the unsatisfiable precondition.
#[kani::requires(a > 5 && a < 4)]
#[kani::ensures(|result: &u32| *result == a)]
fn conjoined_requires(a: u32) -> u32 {
    panic!("This code is never tested")
}

#[kani::proof_for_contract(conjoined_requires)]
fn prove_conjoined_requires() {
    let x: u32 = kani::any();
    conjoined_requires(x);
}
