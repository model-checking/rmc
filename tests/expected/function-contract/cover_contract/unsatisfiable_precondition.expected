contract_cover.1\
	 - Status: FAILURE\
	 - Description: "The contract's precondition is satisfiable."

contract_cover.2\
	 - Status: FAILURE\
	 - Description: "The contract's postcondition is reachable."

VERIFICATION:- FAILED

contract_cover.1\
	 - Status: SUCCESS\
	 - Description: "The contract's precondition is satisfiable."

contract_cover.2\
	 - Status: FAILURE\
	 - Description: "The contract's precondition is satisfiable."

contract_cover.3\
	 - Status: FAILURE\
	 - Description: "The contract's postcondition is reachable."

VERIFICATION:- FAILED
